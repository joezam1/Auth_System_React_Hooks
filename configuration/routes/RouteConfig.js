const routeConfiguration = Object.freeze({
    home:'/',
    authLoginPath:'/auth/login',
    authRegisterPath:'/auth/register',
    authLogoutPath:'/auth/logout',
    privateCustomerDashboard:'/customerDashboard',
    privateCustomerOrders:'/customerOrders',
});

export default routeConfiguration;