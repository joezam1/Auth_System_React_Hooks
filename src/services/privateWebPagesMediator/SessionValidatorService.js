import EnvConfig from '../../../configuration/environment/EnvConfig.js';
import RouteConfig from '../../../configuration/routes/RouteConfig.js';
import ServerConfig from '../../../configuration/server/ServerConfig.js';
import CookieProperties from '../../library/stringLiterals/CookieProperties.js';
import CookieService from '../cookieStorage/CookieService.js';
import LocalStorageService from '../localStorage/LocalStorageService.js';
import inputCommonInspector from '../validators/InputCommonInspector';
import WebWorkerManager from '../../backgroundWorkers/WebWorkerManager.js';
import FetchWorker from '../../backgroundWorkers/FetchWorker.js';
import FetchWorkerHelper from '../../backgroundWorkers/FetchWorkerHelper.js';
import HttpRequestMethods from '../../library/enumerations/HttpRequestMethods.js';
import WindowLocationProperties from '../../library/stringLiterals/WindowLocationProperties.js';
import Helpers from '../../library/common/Helpers.js';




const SessionValidatorService = (function(){
    //Test:DONE
    const redirectPrivateWebpagesMediator = function(redirectTo){
        LocalStorageService.setItemInLocalStorage( WindowLocationProperties.REDIRECT, redirectTo )
        let cookieName = LocalStorageService.getItemFromLocalStorage(CookieProperties.NAME);
        let sessionToken = CookieService.getCookieFromDataStoreByName(cookieName);
        if( !inputCommonInspector.stringIsNullOrEmpty(sessionToken)){
            resolveSessionValidationUsingWebWorkers(FetchWorker, sessionToken);
        }
        else{
            Helpers.setUrlRedirect( RouteConfig.authLogoutPath );
        }
    }

    return {
        redirectPrivateWebpagesMediator : redirectPrivateWebpagesMediator,
    }
})();

export default SessionValidatorService;

//#REGION Private Functions

function resolveSessionValidationUsingWebWorkers(selectedWebWorker, currentSessionToken){
    WebWorkerManager.startNewWorker(selectedWebWorker, validateSessionFetchWorkerMessageCallback);
    let apiRequestMessage = getMessageForApiSession(currentSessionToken);
    WebWorkerManager.sendMessageToWorker(apiRequestMessage);
}


function getMessageForApiSession(cookieValue){
    var sessionUrl = EnvConfig.PROTOCOL +'://' + EnvConfig.TARGET_URL + ServerConfig.apiSessionsSessionTokenGet;
    let requestMethod = HttpRequestMethods[HttpRequestMethods.GET];
    let headersArray = [{name:'x_session_id', value: cookieValue }];
    let payload = {};
    let message = FetchWorkerHelper.getMessageDataForFetchWorker(sessionUrl, requestMethod, headersArray, payload);
    return message;
}


function validateSessionFetchWorkerMessageCallback(event){
    console.log('sessionWorkerOnMessageCallback-event', event)
    let sessionInfo = event?.data?.data?.result;
    WebWorkerManager.terminateActiveWorker()
    if(inputCommonInspector.objectIsValid(sessionInfo)){
        let apiStoredSessionToken = sessionInfo?.sessionToken?.fieldValue;
        let cookieName = LocalStorageService.getItemFromLocalStorage( CookieProperties.NAME);
        let currentSessionCookie = CookieService.getCookieFromDataStoreByName(cookieName);
        if(apiStoredSessionToken === currentSessionCookie){
            let locationRedirect = LocalStorageService.getItemFromLocalStorage(WindowLocationProperties.REDIRECT);
            LocalStorageService.removeItemFromLocalStorage(WindowLocationProperties.REDIRECT);
            if(inputCommonInspector.objectIsValid(locationRedirect)){
                Helpers.setUrlRedirect( locationRedirect );
            }
            return;
        }
    }
    //if is not valid then LOGOUT
    Helpers.setUrlRedirect( RouteConfig.authLogoutPath );
}


//#ENDREGION Private Functions