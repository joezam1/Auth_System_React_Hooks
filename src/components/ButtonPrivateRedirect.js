import React from "react";
import SessionValidatorService from "../services/privateWebPagesMediator/SessionValidatorService";
import InputCommonInspector from '../services/validators/InputCommonInspector.js';
import InputTypeInspector from '../services/validators/InputTypeInspector.js';

export default function ButtonPrivateRedirect(props){

    function executePrivateRedirect(){
        let redirect = props.redirectToLocation;
        if(InputTypeInspector.isTypeString(redirect) && InputCommonInspector.stringIsValid( redirect)){
            SessionValidatorService.redirectPrivateWebpagesMediator( redirect );
        }
    }

    return(<div className='button-private-redirect-section'>
        <button type="button" onClick={()=>{ executePrivateRedirect();  }} data-testid="button-private-redirect-id"> {props.buttonText} </button>
    </div>)
}