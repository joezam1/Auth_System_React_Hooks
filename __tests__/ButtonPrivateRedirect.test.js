import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, cleanup, fireEvent } from '@testing-library/react';

import ButtonPrivateRedirect from "../src/components/ButtonPrivateRedirect";
import SessionValidatorService from "../src/services/privateWebPagesMediator/SessionValidatorService";
import RouteConfig from '../configuration/routes/RouteConfig.js';

jest.mock('../src/services/privateWebPagesMediator/SessionValidatorService.js')


describe('File: ButtonPrivateRedirect.js', function(){
    afterEach(cleanup);
    afterAll(()=>{jest.resetAllMocks(); });

    describe('Component: ButtonPrivateRedirect With Props', function(){

        test('It can take a snapshot', function(){
            let container1 = render( <MemoryRouter> <ButtonPrivateRedirect redirectToLocation={RouteConfig.privateCustomerDashboard} buttonText="test" /> </MemoryRouter>);
            expect(container1.asFragment(<ButtonPrivateRedirect redirectToLocation={RouteConfig.privateCustomerDashboard} buttonText="test"/>)).toMatchSnapshot();
        });


        test('Button Triggers SessionValidatorService', function(){
            //Arrange
            let container2 = render( <MemoryRouter> <ButtonPrivateRedirect redirectToLocation={RouteConfig.privateCustomerDashboard} buttonText="test" /> </MemoryRouter>);
            const button = container2.getByTestId('button-private-redirect-id');
            SessionValidatorService.redirectPrivateWebpagesMediator = jest.fn();
            //Act
            fireEvent.click(button);
            //Assert
            expect( SessionValidatorService.redirectPrivateWebpagesMediator).toHaveBeenCalledTimes(1);
        });
    });

    describe('Component: ButtonPrivateRedirect Without Props', function(){

        test('It can take a snapshot', function(){
            let container3 = render( <MemoryRouter> <ButtonPrivateRedirect /> </MemoryRouter>);
            expect(container3.asFragment(<ButtonPrivateRedirect/>)).toMatchSnapshot();
        });

        test('Button Triggers SessionValidatorService', function(){
            //Arrange
            let container4 = render( <MemoryRouter> <ButtonPrivateRedirect /> </MemoryRouter>);
             const button = container4.getByTestId('button-private-redirect-id');
            SessionValidatorService.redirectPrivateWebpagesMediator = jest.fn();
            //Act
            fireEvent.click(button);
            //Assert
            expect( SessionValidatorService.redirectPrivateWebpagesMediator).toHaveBeenCalledTimes(0);
        });
    });
});
