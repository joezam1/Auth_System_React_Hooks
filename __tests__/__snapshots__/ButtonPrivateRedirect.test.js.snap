// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File: ButtonPrivateRedirect.js Component: ButtonPrivateRedirect With Props It can take a snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="button-private-redirect-section"
  >
    <button
      data-testid="button-private-redirect-id"
      type="button"
    >
       test 
    </button>
  </div>
   
</DocumentFragment>
`;

exports[`File: ButtonPrivateRedirect.js Component: ButtonPrivateRedirect Without Props It can take a snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="button-private-redirect-section"
  >
    <button
      data-testid="button-private-redirect-id"
      type="button"
    >
        
    </button>
  </div>
   
</DocumentFragment>
`;
